File: codex_poting/java/src/main/java/com/example/demo/jooq/service/PurchaseOperationServiceJooq.java
package com.example.demo.jooq.service;

import com.example.demo.api.request.PurchaseInvoiceCreateRequest;
import com.example.demo.api.response.PurchaseInvoiceResponse;
import com.example.demo.config.RetryableOnLockFailure;
import com.example.demo.jooq.converter.PurchaseOperationConverterJooq;
import com.example.demo.jooq.dsl.PurchaseOperationDsl;
import com.example.demo.jooq.entity.PurchaseInvoiceDetailRow;
import com.example.demo.jooq.validator.PurchaseDataValidatorJooq;
import com.example.demo.service.PurchaseOperationService;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;

@Service
@Profile("jooq")
@RequiredArgsConstructor
public class PurchaseOperationServiceJooq implements PurchaseOperationService {
    private final PurchaseOperationDsl purchaseOperationDsl;
    private final PurchaseDataValidatorJooq dataValidator;
    private final PurchaseOperationConverterJooq converter;

    @RetryableOnLockFailure
    @Transactional
    @Override
    public PurchaseInvoiceResponse create(@NonNull PurchaseInvoiceCreateRequest request) {
        final var bookIdsByIsbn = dataValidator.foreignKeyValidate(request);

        final var now = LocalDateTime.now();
        final var amount = converter.calculateAmount(request);
        final var purchaseInvoice = purchaseOperationDsl.insertPurchaseInvoice(request, amount, now);
        final var details = new ArrayList<PurchaseInvoiceDetailRow>();

        request.getDetails().forEach(detailRequest -> {
            final var detailAmount = converter.calculateAmount(detailRequest);
            final var bookId = bookIdsByIsbn.get(detailRequest.getPurchaseInvoiceDetailIsbn());
            final var detail = purchaseOperationDsl.insertPurchaseInvoiceDetail(
                purchaseInvoice.getId(),
                detailRequest,
                bookId,
                detailAmount,
                now
            );
            details.add(detail);
            purchaseOperationDsl.insertBookStockMovement(purchaseInvoice.getReceivingStoreId(), purchaseInvoice, detail, now);
            purchaseOperationDsl.addStockQuantity(purchaseInvoice.getReceivingStoreId(), detail, now);
        });

        return converter.toResponse(purchaseInvoice, details);
    }
}

File: codex_poting/kotlin/src/main/kotlin/com/example/demo/jooq/service/PurchaseOperationServiceJooq.kt
package com.example.demo.jooq.service

import com.example.demo.api.request.PurchaseInvoiceCreateRequest
import com.example.demo.api.request.PurchaseInvoiceDetailCreateRequest
import com.example.demo.api.response.PurchaseInvoiceResponse
import com.example.demo.config.RetryableOnLockFailure
import com.example.demo.jooq.converter.PurchaseOperationConverterJooq
import com.example.demo.jooq.dsl.PurchaseOperationDsl
import com.example.demo.jooq.entity.PurchaseInvoiceDetailRow
import com.example.demo.jooq.validator.PurchaseDataValidatorJooq
import com.example.demo.service.PurchaseOperationService
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

@Service
@Profile("jooq")
class PurchaseOperationServiceJooq(
    private val purchaseOperationDsl: PurchaseOperationDsl,
    private val dataValidator: PurchaseDataValidatorJooq,
    private val converter: PurchaseOperationConverterJooq
) : PurchaseOperationService {
    @RetryableOnLockFailure
    @Transactional
    override fun create(request: PurchaseInvoiceCreateRequest): PurchaseInvoiceResponse? {
        if (request == null) {
            throw NullPointerException("request is marked non-null but is null")
        }
        val bookIdsByIsbn = dataValidator.foreignKeyValidate(request)
        val now = LocalDateTime.now()
        val amount = converter.calculateAmount(request)
        val purchaseInvoice = purchaseOperationDsl.insertPurchaseInvoice(request, amount, now)
        val details = ArrayList<PurchaseInvoiceDetailRow?>()
        request.details!!.filterNotNull().forEach { detailRequest ->
            val detailAmount = converter.calculateAmount(detailRequest)
            val bookId = bookIdsByIsbn.get(detailRequest.purchaseInvoiceDetailIsbn)
            val detail = purchaseOperationDsl.insertPurchaseInvoiceDetail(
                purchaseInvoice.id,
                detailRequest,
                bookId,
                detailAmount,
                now
            )
            details.add(detail)
            purchaseOperationDsl.insertBookStockMovement(
                purchaseInvoice.receivingStoreId,
                purchaseInvoice,
                detail,
                now
            )
            purchaseOperationDsl.addStockQuantity(purchaseInvoice.receivingStoreId, detail, now)
        }
        return converter.toResponse(purchaseInvoice, details)
    }
}
